### PR TITLE
Remove `node-api`

### PR DESCRIPTION
**Pull request by bkchr**
This is the first pull request for https://github.com/paritytech/substrate/issues/750.

There is quite some generic where clause explosion in `node-consensus`, but that will be solved by @gnunicorn as he moves everything into core.

### CODE REVIEW

**Comment by gavofyork** [comment]
`node/api/src` not removed yet?

File: node/consensus/src/service.rs [hunk 1]
```diff
@@ -80,10 +82,17 @@ impl Service {
 		key: ed25519::Pair,
 	) -> Service
 		where
-			A: Api + Send + Sync + 'static,
-			C: BlockchainEvents<Block> + ChainHead<Block> + BlockBody<Block>,
-			C: bft::BlockImport<Block> + bft::Authorities<Block> + Send + Sync + 'static,
-			N: Network + Send + 'static,
+			A: Client + TPClient<Block = <A as Client>::Block> + 'static,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply node/consensus/src/service.rs]: Yeah, that's the `generic where explosion` I spoken about :O 
The crate now is half generic and half typed and that requires all these `where clauses` stuff :(

**Comment by bkchr** [comment]
Yeah, I rebased and the `lib.rs` file sneaked into the code base again.

File: node/consensus/src/lib.rs [hunk 2]
```diff
@@ -73,33 +78,133 @@ pub type SharedOfflineTracker = Arc<RwLock<OfflineTracker>>;
...
+pub trait BlockBuilder<Block: BlockT> {
+	/// Push an extrinsic onto the block. Fails if the extrinsic is invalid.
+	fn push_extrinsic(&mut self, extrinsic: <Block as BlockT>::Extrinsic) -> Result<()>;
+
+	/// Bake the block with provided extrinsics.
+	fn bake(self) -> Result<Block>;
+}
+
+/// Local client abstraction for the consensus.
+pub trait Client: Send + Sync {
```
Comment (arkpar) [inline node/consensus/src/lib.rs]: This should be called `AuthoringApi` or something like that.

File: node/consensus/src/evaluation.rs [hunk 3]
```diff
@@ -56,12 +55,19 @@ error_chain! {
...
 	now: Timestamp,
 	parent_hash: &Hash,
-	parent_number: BlockNumber,
-) -> Result<CheckedBlock> {
+	parent_number: <<Block as BlockT>::Header as HeaderT>::Number,
+) -> Result<CheckedBlock>
+where
+	<<Block as BlockT>::Header as HeaderT>::Number: Add<u64, Output=<<Block as BlockT>::Header as HeaderT>::Number>,
+	<<Block as BlockT>::Header as HeaderT>::Number: PartialEq<u64>,
+	<<Block as BlockT>::Header as HeaderT>::Number: Into<u64>,
```
Comment (arkpar) [inline node/consensus/src/evaluation.rs]: These 3 bounds for `u64` above are not required probably. You can get away with using `As<u64>` which `Number` already supports.

Comment (bkchr) [reply node/consensus/src/evaluation.rs]: Ahh nice, ty. Did not know the `As` trait.

File: node/consensus/src/lib.rs [hunk 4]
```diff
@@ -1,8 +1,10 @@
 // Copyright 2018 Parity Technologies (UK) Ltd.
 // This file is part of Substrate.
 
+
```
Comment (gavofyork) [inline node/consensus/src/lib.rs]: weird newlines crept in...

**Pull request merged**